### ISSUE TITLE
🐛 Bug Report — Runtime APIs: unused SQL cursors lead to strange errors

### ISSUE THREAD

**Issue by mrbbot** [cloudflare/workerd#959]
Hey! 👋 Using the following `workerd` configuration as a template:

```capnp
using Workerd = import "/workerd/workerd.capnp";

const config :Workerd.Config = (
  services = [
    ( name = "main", worker = .worker ),
    ( name = "storage", disk = ( path = "test-storage", writable = true ) ),
  ],
  sockets = [
    ( name = "http", address = "*:8080", http = (), service = "main" ),
  ]
);

const worker :Workerd.Worker = (
  compatibilityDate = "2023-08-01",
  compatibilityFlags = ["experimental"],
  modules = [
    ( name = "index.mjs", esModule = embed "index.mjs" )
  ],
  durableObjectStorage = ( localDisk = "storage" ),
  durableObjectNamespaces = [
    ( className = "DurableObject", uniqueKey = "DurableObject" )
  ],
  bindings = [
    ( name = "OBJECT", durableObjectNamespace = "DurableObject" )
  ],
);
```

```js
// index.mjs
function drain(cursor) {
  for (const _ of cursor) {}
}

export class DurableObject {
  constructor(state) {
    this.state = state;
  }

  async fetch(request) {
    const { pathname } = new URL(request.url);
    if (pathname !== "/") return new Response(null, { status: 404 });

    const sql = this.state.storage.sql;
    sql.exec("DROP TABLE IF EXISTS entries;");
    sql.exec("CREATE TABLE entries (key TEXT PRIMARY KEY, value TEXT);");
    sql.exec("INSERT INTO entries (key, value) VALUES ('key1', 'value1');");

    { /* substitute code from below */ }

    return new Response("OK");
  }
}

export default {
  async fetch(request, env, ctx) {
    const id = env.OBJECT.idFromName("a");
    const stub = env.OBJECT.get(id);
    return stub.fetch(request);
  }
}
```
---

Substituting...

```js
const selectStmt = sql.prepare("SELECT value FROM entries WHERE key = ?");
const selectCursor = selectStmt("key1");
```

...will give `Error: database table is locked` on the 2nd request to the Durable Object, unless `drain(selectCursor)` is called.

---

Similarly, substituting... 

```js
const allCursor = sql.exec("SELECT value FROM entries");
for (const _ of allCursor) break;
```

...to partially read from a cursor, results in the same `Error: database table is locked` on the 2nd request, unless `drain(allCursor)` is also called.

---

Substituting...

```js
const deleteStmt = sql.prepare("DELETE FROM entries WHERE key = ? RETURNING value");
const deleteCursor = deleteStmt("key1");
```

...will give `workerd/util/sqlite.c++:865: failed: SQLite failed; sqlite3_errmsg(db) = cannot commit transaction - SQL statements in progress;  sqlite3_step()` on a request to the Durable Object, unless `drain(deleteCursor)` is called, or `RETURNING value` is removed the statement.

---

I guess it's a bit strange to query data and then not read it all, but it seems like partially reading cursors is something the API permits and encourages. I feel like I'd expect cursors to be dropped at the end of requests, but I guess you could still be using them elsewhere... 😕

**Comment by kentonv** [comment]
Hmm. If you create a new cursor from the _same statement_, we automatically invalidate the previous cursor, because sqlite only wants one copy of a statement running at a time. But we don't invalidate cursor for other statements. Maybe we need to? Or maybe we should be detecting insert/update/delete statements and always running them to completion immediately? (They don't return multiple rows anyway.)

@geelen any thoughts?

**Comment by mrbbot** [comment]
> They don't return multiple rows anyway.

Presumably they could with a [`RETURNING`](https://www.sqlite.org/lang_returning.html) clause? Though I'm not sure if all the updates apply before, or as your iterating through the cursor?

**Comment by geelen** [comment]
Apologies, I missed this when @kentonv originally tagged me. We so far haven't hit this because D1 always reads all cursors to completion, but we should solve this for the general case.

I think the `INSERT ... RETURNING` where you don't read all the rows is the particularly concerning, as you'll get `cannot commit transaction - SQL statements in progress;  sqlite3_step()`. So I think at the end of a request we should automatically clean up all cursors before attempting to commit.

That _should_ (I think) also solve the other cases where subsequent queries fail to run, but if you do this:

```js
const deleteStmt = sql.prepare("DELETE FROM entries WHERE key = ? RETURNING value");
const deleteCursor = deleteStmt("key1");
const allCursor = sql.exec("SELECT value FROM entries");
```

I think you'll get `Error: database table is locked` as well. IMO the only valid use for multiple cursors being open at once is to step through two read queries simultaneously, or stream results from a long query while serving other requests. So I think we can make those possible without leaving footguns like this lying around...

I'll write a test to verify this behaviour and try to find a heuristic that works

**Comment by geelen** [comment]
Right, so the first error appears to be specific to the nature of `DROP TABLE` commands, doing the following explodes:

```js
  sql.exec("DROP TABLE IF EXISTS issue_959;");
  sql.exec("CREATE TABLE issue_959 (key TEXT PRIMARY KEY, value TEXT);");
  sql.exec("INSERT INTO issue_959 (key, value) VALUES ('key1', 'value1'), ('key2', 'value2');");
  sql.exec("SELECT value FROM issue_959 WHERE key = 'key1'"); // unconsumed cursor
  sql.exec("DROP TABLE IF EXISTS issue_959;");
```

with `Error: database table is locked`

Change that last line to an INSERT, not a DROP:

```js
  sql.exec("DROP TABLE IF EXISTS issue_959;");
  sql.exec("CREATE TABLE issue_959 (key TEXT PRIMARY KEY, value TEXT);");
  sql.exec("INSERT INTO issue_959 (key, value) VALUES ('key1', 'value1'), ('key2', 'value2');");
  sql.exec("SELECT value FROM issue_959 WHERE key = 'key1'"); // unconsumed cursor
  sql.exec("INSERT INTO issue_959 (key, value) VALUES ('key3', 'value3'), ('key4', 'value4');");
```

and everything's fine. What's really weird is that you're observing that across subsequent requests, which means that SQLite is doing a COMMIT in between them, but the table isn't being "unlocked"...

I've also confirmed the following:

```js
  sql.exec("DROP TABLE IF EXISTS issue_959;");
  sql.exec("CREATE TABLE issue_959 (key TEXT PRIMARY KEY, value TEXT);");
  sql.exec("INSERT INTO issue_959 (key, value) VALUES ('key1', 'value1'), ('key2', 'value2') RETURNING *;"); // dangling cursor
  sql.exec("SELECT value FROM issue_959 WHERE key = 'key1'"); // dangling cursor
  sql.exec("INSERT INTO issue_959 (key, value) VALUES ('key3', 'value3'), ('key4', 'value4') RETURNING *;"); // dangling cursor
  sql.exec("SELECT value FROM issue_959 WHERE key = 'key4'"); // dangling cursor
```

This fails with `cannot commit transaction - SQL statements in progress;` after the whole block, i.e. only once the Response is trying to be sent. The SELECT statements are returning the data that's just been written just fine. Meaning that, as far as I can observe, cursors from reads and writes can intermingle as long as there are no transaction boundaries, and as long as long as some statements (i.e. DROP TABLE) aren't involved.

To further confirm this, I tried:

```js
  sql.exec("DROP TABLE IF EXISTS issue_959;");
  sql.exec("CREATE TABLE issue_959 (key TEXT PRIMARY KEY, value TEXT);");
  storage.transactionSync(() => {
    sql.exec("INSERT INTO issue_959 (key, value) VALUES ('key1', 'value1'), ('key2', 'value2') RETURNING *;");
  })
  sql.exec("SELECT value FROM issue_959 WHERE key = 'key1'");
  sql.exec("INSERT INTO issue_959 (key, value) VALUES ('key3', 'value3'), ('key4', 'value4') RETURNING *;");
  sql.exec("SELECT value FROM issue_959 WHERE key = 'key4'");
```

And I get `cannot release savepoint - SQL statements in progress;`.

I think the solution here is to automatically consume any open cursors whenever a savepoint/transaction needs to be released. @kentonv @smerritt does that sound right to you?

**Comment by kentonv** [comment]
> I think the solution here is to automatically consume any open cursors whenever a savepoint/transaction needs to be released.

Hmm, does that mean you can't keep a cursor open for streaming? (Since we will commit the automatic transaction as soon as control returns to the event loop.)

**Comment by geelen** [comment]
Yeah I suppose, ideally we'd hook it into some kind of GC logic so that as soon as the JS variable holding the reference to the Cursor falls out of scope it gets cleaned up. If you're streaming you still have a reference to it, so it's fine.

But, in lieu of that, I think we could add a `db.prepare('...').stream()` API that returned a long-lived cursor that bypassed this auto-cleanup operation. That way the default behaviour avoids footguns, and we still support streams but with an explicit API that says things like "You will not be able to delete tables that have an open stream cursor" in the documentation so people know what the downsides are.

**Comment by geelen** [comment]
Pushed some code demonstrating the problem to the `glen/sqlite-unconsumed-cursors` branch: https://github.com/cloudflare/workerd/commit/ee68e52bdd6585826cfa57602993f8254e8db74f

**Comment by kentonv** [comment]
> ideally we'd hook it into some kind of GC logic so that as soon as the JS variable holding the reference to the Cursor falls out of scope it gets cleaned up.

We already do clean up the cursor when it is GC'd. But GC does not proactively apply to things that have gone out-of-scope -- it happens at some arbitrary later point. (Or it might not happen at all, there's no guarantee.)

I think it's a pretty big problem if the success of a new query depends on whether an old cursor has been GC'd yet. At best this leads to unpredictable behavior of apps and subtle footguns. At worst, being able to observe GC might reveal side channels. (This is why we disable WeakRef.)

I think instead, we need the logic to be: If a query cannot be performed while some old cursor still exists, then the old cursor should be implicitly canceled to allow the new query to proceed. If the old cursor is indeed no longer reachable, then this won't be observable to the app. If the app is still holding onto the cursor and tries to use it again, at least they'll get a deterministic error.

> I think we could add a db.prepare('...').stream() API that returned a long-lived cursor that bypassed this auto-cleanup operation.

I think we'd have to go all the way here and create a new DB connection to run the query concurrently. If it's on the same connection, and we have the auto-canceling behavior described above, people may be confused when their stream is broken by some other request coming in.

But if we do create a separate connection, GC behavior is an even bigger concern... if someone forgets to explicitly shut down this connection, it could prevent WAL checkpoints from going forward for arbitrarily long.

Maybe for now we need to tell people: Cursors are meant to be used synchronously. As soon as you return to the event loop, all cursors are invalidated. You will need to build your own higher-level cursor instead.

**Comment by geelen** [comment]
> If a query cannot be performed while some old cursor still exists, then the old cursor should be implicitly canceled to allow the new query to proceed

Is there a way to detect this, rather than just attempting to execute the new query and catching the errors?

> Maybe for now we need to tell people: Cursors are meant to be used synchronously. As soon as you return to the event loop, all cursors are invalidated. You will need to build your own higher-level cursor instead.

That makes sense for cursors returned from writes, since those are currently blocking the transaction commit, though we'd need to remember to also clean things up just before a `transactionSync` (savepoint) block closes.

For reads, I feel like that might be _slightly_ too strict, but probably acceptable. The following code would no longer work, say:

```js
for (row in sql.exec('...')) {
  await env.QUEUES.push({ row })
}
```

Which, while it definitely could be expressed more efficiently, is a pretty natural expression to expect to work. Also, correct me if I'm wrong, but wouldn't the input gate prevent other requests being started while those `QUEUES.push` calls are in progress?

> If it's on the same connection, and we have the auto-canceling behavior described above, people may be confused when their stream is broken by some other request coming in.

Back to this point, I actually don't think we should ever auto-cancel a "currently streaming" cursor. So the fact that you're streaming some data off to the side might prevent you from deleting the table out from under it, but it's kinda up to you to manage things and explicitly close the stream. As long as we give the user a good API, I don't think we need a second connection and to worry about blocking checkpoints.

**Comment by geelen** [comment]
FYI, I just thought I'd double check exactly what an open read cursor _actually_ blocks:

```js
  // DB SETUP
  sql.exec(`CREATE TABLE abc (a INT, b INT, c INT);`)
  sql.exec(`CREATE TABLE cde (c INT, d INT, e INT);`)
  sql.exec(`INSERT INTO abc VALUES (1,2,3),(4,5,6);`)

  // Start a read query then hold the cursor
  const cursor = sql.exec(`SELECT * FROM abc`);
  for (const row of cursor) {
    console.log({before: row})
    break
  }

  // Deleting a table that has an open cursor
  // sql.exec('DROP TABLE IF EXISTS abc;') ❌ Error: database table is locked
  
  // Deleting a different table that does exist
  // sql.exec('DROP TABLE IF EXISTS cde;') ❌ Error: database table is locked

  // All of these work:
  sql.exec('INSERT INTO abc(a,b,c) VALUES (10,11,12),(20,21,22);')
  sql.exec('UPDATE abc SET a = a * 2;')
  sql.exec('DELETE FROM abc where a < 10;')
  sql.exec('DROP TABLE IF EXISTS xyz;')
  sql.exec('CREATE TABLE xyz (x INT, y INT, z INT);')

  for (const row of cursor) {
    console.log({after: row})
  }
```

This outputs:

```
workerd/io/worker.c++:1661: info: console.log(); message() = [{"before":{"a":1,"b":2,"c":3}}]
workerd/io/worker.c++:1661: info: console.log(); message() = [{"after":{"a":20,"b":11,"c":12}}]
workerd/io/worker.c++:1661: info: console.log(); message() = [{"after":{"a":40,"b":21,"c":22}}]
```

Which is kinda wild. It's reading rows that were inserted and skipping over rows that were deleted _after_ the cursor was created.

I couldn't find any other statements (apart from dropping tables) that would actually fail, or make the cursor return nonsense.

**Comment by smerritt** [comment]
The weird behavior is because we're running multiple statements concurrently on a single database connection. 

From https://www.sqlite.org/isolation.html: "SQLite provides isolation between operations in separate database connections. However, there is no isolation between operations that occur within the same database connection."

**Comment by kentonv** [comment]
> Is there a way to detect this, rather than just attempting to execute the new query and catching the errors?

Dunno, but that was what I was thinking we'd do: If a query returns an error indicating that it can't run because of another query, we then go cancel all other queries. Big hammer but I think it gets the job done.

> Back to this point, I actually don't think we should ever auto-cancel a "currently streaming" cursor.

There are three options:
1. New query auto-cancels conflicting queries.
2. New query fails until conflicting query is closed or GC'd.
3. New query fails until conflicting query is explicitly closed. GC is ignored -- cursors stay open even if they are no longer reachable from JS.

I feel pretty strongly that (2) is not an acceptable choice here. GC is non-deterministic and we really do not want non-deterministic things to be visible to apps. In addition to potential side channel risks, this would make it hard for us to build record/replay, where the exact state of an isolate can be reproduced by replaying recorded inputs.

Between (1) and (3), it seems like (1) is the lesser evil. If we went with (3), that implies that if someone _ever_ forgets to close a cursor, their connection could be stuck in a broken state until the DO resets.

**Comment by smerritt** [comment]
As far as unused cursors go, I think we could get some help by having the cursor eagerly consume the first result row from the database, saving it in case the application wants it later.

Lots of inserts, updates, and deletes return a single result row which is just the number of DB rows affected. Buffering one row would consume all the results, so SQLite no longer cares about it.

It wouldn't fix everything, e.g. `UPDATE ... RETURNING` would still need to be consumed, but it might help.

**Comment by geelen** [comment]
I like that idea @smerritt. Additionally

> There are three options:

I don't think the solution falls neatly into one of these three options. What I propose is this:

* All cursors from a write get automatically canceled directly before a savepoint is released/transaction is committed.
* A new `const dispose = cursor.hold()` function is introduced, allowing a user to explicitly mark a cursor as "not for automatic cancellation"
  * `dispose` here is a function that cancels the cursor. No need for GC.
* New queries auto-cancel conflicting cursors, except those marked as "held".
* A new `storage.sql.cancelAllCursors()` function is added which cancels everything, regardless of whether or not it's "held". This can be used as an escape hatch to avoid needing to fully reset the object.
 
Note that so long as write cursors are auto-cancelled, the only thing that long-lived read cursors appear to block is deleting tables. So "auto-cancel conflicting queries" won't often be invoked.

That said, the very _presence_ of the explicit `cursor.hold()` and `cancelAllCursors()` (names TBB 🚲🏠) gives us a chance to document, explain & warn users about how cursors actually work under the hood, but allows them the power to build long-running streaming APIs. But it means we don't rely on GC or a "big hammer" that ends up limiting their options.

**Comment by kentonv** [comment]
@smerritt As it happens we already do step the first row of the query before returning the cursor. But I guess what we really have to do is perform a _second_ step such that sqlite has a chance to tell us the query is done, if it turns out it only returns one row. Yeah, I think that's reasonable.

@geelen Hmm. I think by the time you have a cursor, it's too late to call something like `hold()` on it, because we have to make sure such cursors aren't part of an auto-transaction. You probably have to specify upfront that you want a long-lived cursor.

Does an open read cursor (not on its own connection) block checkpointing?

**Comment by geelen** [comment]
> I think by the time you have a cursor, it's too late to call something like hold() on it, because we have to make sure such cursors aren't part of an auto-transaction.

I don't understand why this would be the case? It would be synchronous, called directly after the cursor is created, so I don't think there's a transaction boundary at all?

> Does an open read cursor (not on its own connection) block checkpointing?

Sadly, yes. I just tested it. If I add `wal_autocheckpoint = 1` and watch the `test_tmpdir` as the test is running I can see the WAL growing then shrinking. As soon as I hold an open read cursor, it's clearly not checkpointing.

> Maybe for now we need to tell people: Cursors are meant to be used synchronously. As soon as you return to the event loop, all cursors are invalidated. You will need to build your own higher-level cursor instead.

This is starting to sound like a better short-term solution. Perhaps the cursors we have right now should be 100% synchronous, and if we decide to offer long-lived, streaming queries, we do it with something totally different like: `sql.prepare().stream(timeout, async (cursor) => ...)`.

**Comment by kentonv** [comment]
> I don't understand why this would be the case? It would be synchronous, called directly after the cursor is created, so I don't think there's a transaction boundary at all?

I mean that if a previous write query caused an auto-transaction to be started, then the cursor is going to be part of the transaction, which will end as soon as we return to the event loop. If a read cursor was created during a transaction, can it outlive the transaction?

> Perhaps the cursors we have right now should be 100% synchronous, and if we decide to offer long-lived, streaming queries, we do it with something totally different like:

Yeah, that sounds right to me.

How important is streaming for D1?

**Comment by geelen** [comment]
> If a read cursor was created during a transaction, can it outlive the transaction?

Yep, I just tried:

```js
  sql.exec(`DROP TABLE IF EXISTS ab`)
  sql.exec(`CREATE TABLE ab (a INT, b INT);`)
  sql.exec(`INSERT INTO ab VALUES (1,2),(3,4),(5,6);`)

  const cursor = sql.exec(`SELECT * FROM ab where a `);
  for (const row of cursor) {
    console.log({before: row})
    break
  }

  await scheduler.wait(1)

  for (const row of cursor) {
    console.log({after: row})
  }
```

that should do it, right? Console logs out:

```
workerd/io/worker.c++:1661: info: console.log(); message() = [{"before":{"a":1,"b":2}}]
workerd/io/worker.c++:1661: info: console.log(); message() = [{"after":{"a":3,"b":4}}]
workerd/io/worker.c++:1661: info: console.log(); message() = [{"after":{"a":5,"b":6}}]
```

> How important is streaming for D1?

We do have some users running out of memory trying to buffer the results of large queries, but we also haven't done any work yet to optimise that process, so it's hard to say how much we really need a new API. We can definitely continue to make things work without it, folks might just need to do their own pagination/cursors. @elithrar might have more context on this tho.
